File: backend/app/schemas/shared_reporting.py
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from typing import Literal
import uuid

from app.schemas import BaseSchema


ReportingModuleKey = Literal["restaurant_pos", "takeaway_pos", "retail_pos"]
ReportingFreshnessStatus = Literal["disabled", "no_data", "current", "stale", "degraded"]


class SharedSalesMetrics(BaseSchema):
    order_count: int
    void_count: int
    refund_count: int
    gross_sales: Decimal
    discount_amount: Decimal
    tax_amount: Decimal
    refund_amount: Decimal
    net_sales: Decimal


class SharedSalesModuleSummary(SharedSalesMetrics):
    module_key: ReportingModuleKey


class SharedSalesWorkspaceSummary(SharedSalesMetrics):
    module_key: ReportingModuleKey
    brand_id: uuid.UUID
    brand_name: str
    branch_id: uuid.UUID
    branch_name: str


class SharedSalesDailySummary(SharedSalesMetrics):
    business_date: date


class SharedSalesDocumentRead(BaseSchema):
    module_key: ReportingModuleKey
    brand_id: uuid.UUID
    brand_name: str
    branch_id: uuid.UUID
    branch_name: str
    business_date: date
    source_document_type: str
    source_document_id: uuid.UUID
    document_number: str | None
    source_status: str
    gross_sales: Decimal
    refund_amount: Decimal
    net_sales: Decimal
    entry_route: str


class SharedReportingFreshness(BaseSchema):
    status: ReportingFreshnessStatus
    projector_enabled: bool
    projection_mode: Literal["shadow"] = "shadow"
    is_source_of_truth: bool = False
    last_projected_at: datetime | None
    last_polled_at: datetime | None
    lag_seconds: int | None
    failed_sources: list[str]


class SharedSalesReportRead(BaseSchema):
    company_id: uuid.UUID
    date_from: date
    date_to: date
    generated_at: datetime
    freshness: SharedReportingFreshness
    totals: SharedSalesMetrics
    modules: list[SharedSalesModuleSummary]
    workspaces: list[SharedSalesWorkspaceSummary]
    daily: list[SharedSalesDailySummary]
    recent_documents: list[SharedSalesDocumentRead]

File: frontend/src/types/sharedReporting.ts
export type ReportingModuleKey = "restaurant_pos" | "takeaway_pos" | "retail_pos";

export type MoneyValue = number | string;

export type SharedSalesMetrics = {
  order_count: number;
  void_count: number;
  refund_count: number;
  gross_sales: MoneyValue;
  discount_amount: MoneyValue;
  tax_amount: MoneyValue;
  refund_amount: MoneyValue;
  net_sales: MoneyValue;
};

export type SharedSalesModuleSummary = SharedSalesMetrics & {
  module_key: ReportingModuleKey;
};

export type SharedSalesWorkspaceSummary = SharedSalesMetrics & {
  module_key: ReportingModuleKey;
  brand_id: string;
  brand_name: string;
  branch_id: string;
  branch_name: string;
};

export type SharedSalesDailySummary = SharedSalesMetrics & {
  business_date: string;
};

export type SharedSalesDocument = {
  module_key: ReportingModuleKey;
  brand_id: string;
  brand_name: string;
  branch_id: string;
  branch_name: string;
  business_date: string;
  source_document_type: string;
  source_document_id: string;
  document_number: string | null;
  source_status: string;
  gross_sales: MoneyValue;
  refund_amount: MoneyValue;
  net_sales: MoneyValue;
  entry_route: string;
};

export type SharedReportingFreshness = {
  status: "disabled" | "no_data" | "current" | "stale" | "degraded";
  projector_enabled: boolean;
  projection_mode: "shadow";
  is_source_of_truth: false;
  last_projected_at: string | null;
  last_polled_at: string | null;
  lag_seconds: number | null;
  failed_sources: string[];
};

export type SharedSalesReport = {
  company_id: string;
  date_from: string;
  date_to: string;
  generated_at: string;
  freshness: SharedReportingFreshness;
  totals: SharedSalesMetrics;
  modules: SharedSalesModuleSummary[];
  workspaces: SharedSalesWorkspaceSummary[];
  daily: SharedSalesDailySummary[];
  recent_documents: SharedSalesDocument[];
};

export type SharedSalesReportFilters = {
  date_from: string;
  date_to: string;
  module_key?: ReportingModuleKey;
  brand_id?: string;
  branch_id?: string;
};
